Allow defining translation functions in a module for reuse. #9

0xAX opened the pull request.

bitwalker commented:
@0xAX Could you add a test + documentation on how to use it? Looks pretty good to me, just one question though: have you tested this within the context of a release with exrm? I don't know if you are using exrm, so if not, no worries.

liveforeverx commented:
@bitwalker referenced #9, it is implemented with exrm part ( bitwalker/exrm#160 ).

liveforeverx commented on lib/conform.ex:
```diff
@@ -73,10 +73,30 @@ defmodule Conform do
 
+  defp load_imports(opts) do
+    release_dir = Path.dirname(opts.schema)
+    {:ok, [_ | zip_files]} = :zip.list_dir(("#{release_dir}/app.schema.deps.ez") |> to_char_list)
```
In conjunction, if there no imports and no archive created (  https://github.com/bitwalker/exrm/pull/160/files#diff-3af356a8acb8942aeae5ea2692f69360R48 ), what happens on release start with conform configuration, without archive, but assuming that it is always here?

bitwalker replied:
Good catch, this should definitely gracefully handle the case where the zip doesn't exist.

0xAX commented:
Hello @bitwalker, I've updated PR.

Changes since previous version: Added documentation to the README.md and creation of archive fixed. Tests for this are in the exrm - https://github.com/bitwalker/exrm/pull/160

bitwalker commented:
@0xAX Awesome! Do we have any documentation/tests on the imports feature in conform itself? It may be rare, but I'd like to make sure it's covered outside the context of exrm. The current documentation you've added to the README implies it can only be used with releases, but it seems to me that it's more general than that, and in fact the behavior with exrm is/should be transparent to the end user. Thoughts?

0xAX commented:
@bitwalker, 

> Do we have any documentation/tests on the imports feature in conform itself

Did you mean ability to use `import` feature without `exrm`? We have no this feature for now, because conform code uses archive generated by the exrm now.

> and in fact the behavior with exrm is/should be transparent to the end user

What did you mean here?

cc: @liveforeverx

bitwalker commented:
@0xAX Ah ok, I guess that's where my confusion lies then, as I was assuming the import feature worked both inside and outside of releases. Is this something we should consider? I don't mind merging this for the time being and making that a separate thing, for either yourself or someone else to work on.

Since the feature is currently specific to releases, you can ignore my comment from your second quote :)

liveforeverx commented:
@bitwalker @0xAX I think there should be documented, that it is possible (for example in mix conform.effective) use any included module as it happens after `loadpaths` tasks, but in case, if conform script will be generated, there is only included conform with elixir themself.

If somebody want to include any another modules for use, he should add .ez archive with needed applications and modules in it alongside of schema. This has definitely use without exrm too. We can may be add task mix conform.package, which will be generate the archive as exrm do, and change exrm to simple use the function, which generates this archive.

liveforeverx commented:
@bitwalker there is no tasks to generate full .schema in conform themself, ok than we don't need conform.package.

liveforeverx commented:
@0xAX May be add `mix conform.gen_schema`, which doing something (what mix conform.effective doesn't do, but exrm does it) like merge all schemas(what happens in exrm packaging) from sub applications and generate archive from import. After that, it can be used definitely without exrm. Without any release tool or with any other release tool. @bitwalker ?

0xAX commented:
@bitwalker I've updated conform and exrm. Now conform provided mix conform.archive task to create an archive that contains dependencies pointed by the `import` in a schema. So this feature can be used as in exrm so in conform without exrm.

bitwalker commented:
Awesome, I think this is good to merge!

@liveforeverx any last thoughts?

bitwalker commented:
Thanks @0xAX :)
